src/devices/src/virtio/block/device.rs

```diff
@@ -0,0 +1,285 @@
+// for `VirtioDevice`.
+impl<M: GuestAddressSpace + Clone + Send + 'static> WithVirtioConfig<M> for Block<M> {
+    fn device_type(&self) -> u32 {
+        BLOCK_DEVICE_ID
+    }
+
+    fn virtio_config(&self) -> &VirtioConfig<M> {
+        &self.virtio_cfg
+    }
+
+    fn virtio_config_mut(&mut self) -> &mut VirtioConfig<M> {
+        &mut self.virtio_cfg
+    }
+}
+
+impl<M: GuestAddressSpace + Clone + Send + 'static> WithDeviceOps for Block<M> {
+    type E = Error;
+
+    fn activate(&mut self) -> Result<()> {
+        if self.virtio_cfg.device_activated {
+            return Err(Error::AlreadyActivated);
+        }
+
+        if !self.queues_valid() {
+            return Err(Error::QueuesNotValid);
+        }
+
+        // We do not support legacy drivers.
+        if self.virtio_cfg.driver_features & (1 << VIRTIO_F_VERSION_1) == 0 {
+            return Err(Error::BadFeatures(self.virtio_cfg.driver_features));
+        }
+
+        // Set the appropriate queue configuration flag if the `EVENT_IDX` features has been
+        // negotiated.
+        if self.virtio_cfg.driver_features & (1 << VIRTIO_F_RING_EVENT_IDX) != 0 {
+            self.virtio_cfg.queues[0].set_event_idx(true);
+        }
+
+        let ioeventfd = EventFd::new(EFD_NONBLOCK).map_err(Error::EventFd)?;
+
+        // Register the queue event fd.
+        self.vm_fd
+            .register_ioevent(
+                &ioeventfd,
+                &IoEventAddress::Mmio(
+                    self.mmio_cfg.range.base().0 + VIRTIO_MMIO_QUEUE_NOTIFY_OFFSET,
+                ),
+                0u32,
+            )
+            .map_err(Error::RegisterIoevent)?;
+
+        let file = OpenOptions::new()
+            .read(true)
+            .write(!self.read_only)
+            .open(&self.file_path)
+            .map_err(Error::OpenFile)?;
+
+        let mut features = self.virtio_cfg.driver_features;
+        if self.read_only {
+            // Not sure if the driver is expected to explicitly acknowledge the `RO` feature,
+            // so adding it explicitly here when present just in case.
+            features |= 1 << VIRTIO_BLK_F_RO;
+        }
+
+        // TODO: Create the backend earlier (as part of `Block::new`)?
+        let disk = StdIoBackend::new(file, features).map_err(Error::Backend)?;
+
+        let driver_notify = SingleFdSignalQueue {
+            irqfd: self.irqfd.clone(),
+            interrupt_status: self.virtio_cfg.interrupt_status.clone(),
+        };
+
+        let inner = InOrderQueueHandler {
+            driver_notify,
+            queue: self.virtio_cfg.queues[0].clone(),
+            disk,
+        };
+
+        let handler = Arc::new(Mutex::new(QueueHandler { inner, ioeventfd }));
+
+        // Register the queue handler with the `EventManager`. We could record the `sub_id`
+        // (and/or keep a handler clone) for further interaction (i.e. to remove the subscriber at
+        // a later time, retrieve state, etc).
+        let _sub_id = self
```

Comment:
Should we remember this subscriber id and actually use it in the process function to de-register the device when we encounter an error?